### DISCUSSION TITLE
Audio 2ch mic example

### DISCUSSION THREAD

**Discussion by kaspernyhus** [hathach/tinyusb#1711] [Q&A]
Please see: [audio_2_channel_mic_freertos](https://github.com/kaspernyhus/tinyusb/tree/audio_2_channel_mic_freertos/examples/device/audio_2_channel_mic_freertos)

I'm trying to make a 2 channel microphone example building on the 4 channel mic example I ported to freertos to run on the Espressif ESP32-S3. This example outputs 4 triangle waves at different periods and is working (image attatched)
![Screenshot from 2022-11-01 13-55-43](https://user-images.githubusercontent.com/46648238/199237137-64223fb5-59b8-4ce7-ad82-f7f3855d4efe.png)

However now when I'm trying to do the same with 2 channels, I don't see any payload data in the USB frames.

I have made new descriptors to accommodate for the change in channel count, but apart from that it should be the same. I'm not using encoding in both 4ch and 2ch examples, and therefor use `tud_audio_write` to transfer the sample bytes to the USB FIFO. I'm pretty sure the device is enumerated correctly as it shows up and Linux doesn't complain about it. Wireshark reveals that the host is actually requesting the 192 bytes pr frame as expected, but the device returns 0 bytes.

@PanRe do you have an idea about why this could be?

Many thanks in advance!

Kasper Nyhus

**Comment by PanRe** [comment]
Difficult to say from here. The driver usually sends what is stored in its `ep_in_ff` FIFO once a transmit command was received. Have a look at this [line](https://github.com/hathach/tinyusb/blob/28f49c088bb0c498d730d80943017172061cfd05/src/class/audio/audio_device.c#L912) via a breakpoint and look how many bytes are stored in the FIFO! Let me know what it reports!

**Comment by kaspernyhus** [comment]
OK, I have had a look at this and I find that the correct number of bytes are present in the buffers. The esp32 has USE_LINEAR_BUFFER_TX set as `dcd_edpt_xfer_fifo` in the device driver [is not supported](https://github.com/hathach/tinyusb/blob/master/src/portable/espressif/esp32sx/dcd_esp32sx.c#L55)


For 1 channel there are 96bytes in both `ep_in_ff.buffer` and `lin_buf_in`
![1ch](https://user-images.githubusercontent.com/46648238/199942250-b38632f8-032a-43cc-b729-6c8abc0b7b36.png)

and for 2 channels there are 192 bytes
![2ch](https://user-images.githubusercontent.com/46648238/199943923-78a42b47-0e45-4d6e-91f5-4a3cb8a89c10.png)

all as expected

What does this suggest?

Thanks for your help!

**Reply by PanRe** [reply]
This tells us, there is data available but not sent ;)

Ok, could you please verify, that the function `usbd_edpt_xfer()` is acutally called and returns true?
Also, please verify, that the ISO EP was opened successfully. You may debug via breakpoint [here](https://github.com/hathach/tinyusb/blob/28f49c088bb0c498d730d80943017172061cfd05/src/class/audio/audio_device.c#L1611).

**Reply by PanRe** [reply]
You also check if `audiod_open()` returns successfully. This function verifies your descriptors.

**Reply by kaspernyhus** [reply]
Right, so far so good :)

Looking in to this, I don't think I've found where it goes wrong.

`audiod_open()` called from [here](https://github.com/hathach/tinyusb/blob/master/src/device/usbd.c#L907) returns `drv_len` of 128 (1ch version returns 124 by comparison)
![2ch audiod_open](https://user-images.githubusercontent.com/46648238/200139271-a9a87894-a66a-4b56-b3fc-6e0f62e38d9a.png)
so descriptors should be valid?

`usbd_edpt_open()` returns true
![2ch usbd_edpt_open ret true](https://user-images.githubusercontent.com/46648238/200139633-aa338f90-4dad-42ed-880e-6b15bcfc37ac.png)

and `usbd_edpt_xfer()` called [here](https://github.com/hathach/tinyusb/blob/28f49c088bb0c498d730d80943017172061cfd05/src/class/audio/audio_device.c#L909) also returns true
![2ch usbd_edpt_xfer ret true](https://user-images.githubusercontent.com/46648238/200140229-3d22d1f8-c348-4929-880e-debde4a426f6.png)

After enumeration completes and the ISO IN transfers start the host is requesting the 192 bytes but receives 0
![2ch_ws_host](https://user-images.githubusercontent.com/46648238/200140982-7fea2f8f-da62-48c5-b949-d6f3c14cc8c3.png)
![2ch_ws_device](https://user-images.githubusercontent.com/46648238/200140985-c7c40b3e-bd03-4b20-a6ec-0c5a361a7581.png)

**Reply by kaspernyhus** [reply]
by comparison 1ch version wireshark
![1ch_ws_device](https://user-images.githubusercontent.com/46648238/200141164-3adb9cdf-acbe-4b26-9f91-28728a59a662.png)

**Reply by kaspernyhus** [reply]
Could this be a bug in the Espressif device driver?

**Reply by PanRe** [reply]
Do you use the PR #1690? It was merged some time ago fixing a bug for ESP32-S2/S3.

**Reply by kaspernyhus** [reply]
> Do you use the PR #1690? It was merged some time ago fixing a bug for ESP32-S2/S3.

Yes
```
kaspernyhus@HP-ZBook:~/esp/usb_audio_test/components/tinyusb_audio/tinyusb (master)$ git log
commit 28f49c088bb0c498d730d80943017172061cfd05 (HEAD -> master, origin/master, origin/HEAD)
Merge: 5b1b383d2 815c2cc99
Author: Ha Thach <thach@tinyusb.org>
Date:   Tue Oct 25 15:23:09 2022 +0700

    Merge pull request #1698 from battlesnake/fix-attrs-on-struct-member
    
    do not apply storage attributes to member of struct

commit 815c2cc9959cc9a127d4ff64f65743d4ac8713ce
Author: Mark K Cowan <mark@battlesnake.co.uk>
Date:   Mon Oct 24 20:36:45 2022 +0300

    keep alignment spec

commit 5b1b383d2a1eb54ab7f954795d48fbc882de9634
Merge: bfcdef499 e3b7ed9ae
Author: Ha Thach <thach@tinyusb.org>
Date:   Mon Oct 24 19:08:58 2022 +0700

    Merge pull request #1676 from tswan-quasi/master
    
    dcd_lpc_ip3511: isochronous support and endpoint accidental write fix

commit e3b7ed9ae966be1cb41165343ef30c6523ed5771
Author: hathach <thach@tinyusb.org>
Date:   Mon Oct 24 18:43:21 2022 +0700

    use dummy for all ZLP for ip3511, fix lpc55 build with DEBUG=1

commit 35e1a27c950c4b25106c074d0b66220d099e54a7
Author: tswan-quasi <tucker.swan@10xar.com>
Date:   Fri Oct 21 15:06:30 2022 -0400

    unused (void) cast

commit bfcdef499a2b2f7b85047216cea74354b32f2d06
Merge: 55db123a8 a49ca795a
Author: Ha Thach <thach@tinyusb.org>
Date:   Sat Oct 22 00:34:38 2022 +0700

    Merge pull request #1690 from tore-espressif/fix/espressif/iso_transfers
    
    esp: Fix Isochronous transfers

```
I was the one who opened issue [#1382](https://github.com/hathach/tinyusb/issues/1382) back in March :)

**Reply by PanRe** [reply]
Ah yes, sorry for the silly question. Well, yes i guess the audio driver is not to blame here... There is data and it trys to send everything.
@tore-espressif could you assist here once more please?

**Reply by kaspernyhus** [reply]
Turns out switching to another device driver fixes the problems I have.
See [here](https://github.com/leeebo/tinyusb_src_esp/blob/master/Kconfig#L3)

**Reply by PanRe** [reply]
Interesting... well i am not aware of all the device drivers out there.

@hathach how should we handle all the device drivers? I guess fixing bugs in dwc2 and the individual device drivers is not worth it. Shouldn't all drivers which are covered by dwc2 put away?

**Reply by hathach** [reply]
@PanRe yeah, the goal is using dwc2 for all chips. However, since lots of espressif codebase is already using dcd_esp32sx.c and they continue to make pr/fix to that driver. They are probably more conformable with esp specific driver. I guess we will do syncing driver one after a while. And giving user option to choose from since the generic dwc2 is far from perfect now. I will do cherry pick fix to dwc2 later on (current buried in other works)

**Reply by tore-espressif** [reply]
Hello @hathach , I'd like to shed some more light into our plans with DCD layer.
Although we made some hotfixes to the esp32sx DCD, we would like to move the dwc2 implementation. It is already working in our setups, we just need to do more testing before opening PR in your upstream project. We'll keep you updated!

**Reply by hathach** [reply]
@tore-espressif awesome thank you for your update, that is great to hear you plan to move to dwc2 as well. The driver has more code and probably a bit more complicated since it support a wider range of synopsys configuration, especially those with highspeed PHY support. When I have more time to work on dwc, I will migrate/upgrade/update hotfix from dcd_esp32sx to dwc2 with more testing as well.

**Comment by StuartIanNaylor** [comment]
https://github.com/ArmDeveloperEcosystem/microphone-library-for-pico/issues/3#issuecomment-1067540589 supposedly 4 mic has been done

**Reply by kaspernyhus** [reply]
Are you aware of any 2 channel mic implementations?
Thanks a lot

**Reply by StuartIanNaylor** [reply]
I am presuming you can just hack the above as with UAC audio you get chan1,chan2,chan3,chan4 in sequence with stereo its just chan1,chan2.
Only had a brief look but think they used tinyusb as seen a few for a single mic but thinking the above should give you a gist on how to handle channels.

**Reply by kaspernyhus** [reply]
Also what I thought. I can do 1 channel and 4 channels successfully but I’m not able do a stereo mic

**Reply by delingren** [reply]
Have you ever found a 2-channel implementation? I'm trying to do exactly the same thing. I got the device recognized as a stereo mic but the data is messed up. I'm not super well versed in USB protocol. It'd be nice to be able to copy an implementation or use it as a reference.
Thanks.
